src/nvim/option.c

```diff
@@ -2197,6 +2197,53 @@ static const char *did_set_laststatus(optset_T *args)
   return NULL;
 }
 
+/// Process the updated 'lines' or 'columns' option value.
+static const char *did_set_lines_or_columns(optset_T *args)
+{
+  // If the screen (shell) height has been changed, assume it is the
+  // physical screenheight.
+  if (p_lines != Rows || p_columns != Columns) {
+    // Changing the screen size is not allowed while updating the screen.
+    if (updating_screen) {
+      OptVal oldval = (OptVal){ .type = kOptValTypeNumber, .data = args->os_oldval };
+      set_option_varp(args->os_idx, args->os_varp, oldval, false);
+    } else if (full_screen) {
+      screen_resize((int)p_columns, (int)p_lines);
+    } else {
+      // TODO(bfredl): is this branch ever needed?
+      // Postpone the resizing; check the size and cmdline position for
+      // messages.
+      Rows = (int)p_lines;
+      Columns = (int)p_columns;
+      check_screensize();
+      int new_row = (int)(Rows - MAX(p_ch, 1));
+      if (cmdline_row > new_row && Rows > p_ch) {
+        assert(p_ch >= 0 && new_row <= INT_MAX);
+        cmdline_row = new_row;
+      }
+    }
+    if (p_window >= Rows || !option_was_set(kOptWindow)) {
+      p_window = Rows - 1;
+    }
+  }
+
+  // Adjust 'scrolljump' if needed.
+  if (p_sj >= Rows && full_screen) {
+    p_sj = Rows / 2;
+  }
+
+  // If 'scroll' became invalid because of a side effect silently adjust it.
+  if (full_screen) {
+    if (curwin->w_p_scr <= 0) {
+      curwin->w_p_scr = 1;
+    } else if (curwin->w_p_scr > curwin->w_height_inner && curwin->w_height_inner > 0) {
+      curwin->w_p_scr = curwin->w_height_inner;
+    }
+  }
```

Comment:
'lines' and 'columns' aren't the only options that can change window height. 'winheight' and 'laststatus' can as well.